proposals/bip-003.md

```diff
@@ -0,0 +1,460 @@
+# Agglutinative Roles language
+
+| Field | Value |
+| --- | --- |
+| BIP | 3 |
+| Title | Agglutinative Roles language |
+| Author | Alfredo Beaumont |
+| Status | Draft |
+| Created | 2017-09-04 |
+| Updated | 2017-09-04 |
+
+## Abstract
+
+Current Role set includes roles that contain multiple properties,
+instead of having a single Role per property.
+
+We propose modifying current role set to include only roles that represent a single property
+and to consider the Role set an agglutinative language,
+annotating each node with a set of Roles to fully express node properties.
+
+We show how this can make the Role vocabulary more scalable, flexible and expressive,
+and how it can make the work of code analysts easier.
+
+## Rationale
+
+In current Role set many roles contain multiple properties (i. e.: `OpPreIncrement` for `++i`).
+
+This presents some issues:
+* It doesn't scale well.
+  For a set of N properties, in the worst case 2^N roles would be needed.
+* It makes code analysis hard.
+  For example, to search for all the operators you need to search for the union of many roles.
+* There's hard to provide partial information for rare properties,
+  if some property is not common enough as to deserve its own role.
+
+A more flexible and scalable way to define the role set would be to have one role per property
+and to allow to combine these roles/properties as needed.
+In the example above `OpPreIncrement` would become the union of 3 roles (`Operator`, `Increment`, `Prefix`).
+
+This combination of roles is already done to some extent,
+since a preincrement operator would actually be annotated with 2 roles: `Expression`, `OpPreIncrement`.
+Current proposal just deepens this property separation into roles.
+
+As a derivative advantage of this new role language and its scalability,
+new roles could be defined that better capture some additional property shared by different tokens
+that will make code analysis even easier.
+
+Following the previous example, we could use some additional roles, like `Arithmetic`, `Unary`, etc.
+to express additional properties of the role.
+So the final annotation could be something like: `Operator`, `Arithmetic`, `Unary`
+
+* `Expression` (or `Statement`, depending on the language)
+* `Operator`
+* `Arithmetic`
+* `Increment`
+* `Unary`
+* `Prefix`
+
+With this kind of classification a code analyst that may be interested in analysing the usage of, say,
+arithmetic operators, would have an easier way to filter the `UAST` to find the interesting nodes.
+
+Additionally, this agglutination of roles makes unsupported node types degrade more gracefully.
+For example, a `log` operator, which currently lacks an specific role,
```

Comment:
Nitpicking: at first I tough about logging, maybe change for `natural logarithm` operator?

Reply:
lacks *a* specific

Reply:
Changed to `**` (pow), to avoid potential confusion.